### ISSUE TITLE
Transposition of fretboard diagrams and chord symbols

### ISSUE THREAD

**Issue by Dima-S-Jr** [musescore/MuseScore#23712]
## My idea

I propose to implement automatic modification of the fretboard diagram in accordance with the key into which the user can optionally transpose the score.

## Problem to be solved

Currently, when transposing the score to any other key, the fretboard diagrams do not change (that is, they remain in the same key in which they were originally composed). Therefore, the user will have to manually redo each fretboard diagram if they suddenly need to change the key. I suggest avoiding such fuss by allowing the program itself to automatically change the fretboard diagram in accordance with the key set by the user. This will reduce the time spent writing the score.

## Additional context

I also ask you to pay attention to the fact that when changing the key of the score, not only the fretboard diagrams do not change, but also the alphanumeric notation of the chords that are above the fretboard diagrams.

https://github.com/user-attachments/assets/aadd4603-6d3e-4336-a241-181dfebccff0

## Checklist

- ✅ This request follows the [guidelines](https://github.com/musescore/MuseScore/wiki/Reporting-bugs-and-issues) for reporting issues
- ✅ I have verified that this feature request has not been logged before, by searching the [issue tracker](https://github.com/musescore/MuseScore/issues) for similar requests

**Comment by cbjeukendrup** [comment]
My first thought: I wonder whether it is even theoretically possible to design an algorithm for this...

On the other hand, maybe it is doable, if you approach it in the following way. In a way, this problem is not very different from the problem of distributing staff notation notes over the strings of a TAB staff, which we can apparently already do because we can create linked staves. 

So an algorithm could be:
1. convert from fret diagram to a list of pitches
2. transpose those pitches using the pitch transposition logic that we already have
3. convert the transposed pitches back to strings+frets, using the logic that we already have for linked staves
4. put those strings+frets into a fret diagram

That will probably fail to handle fingerings, barrés and whatever other fancy things a fret diagram might have, but perhaps those can be handled separately in some way. 

Anyway, there will inevitably be cases where human intervention will be required, but if we can build a system that comes up with something nice in 80% of the cases, that's already nice too.

**Comment by MarcSabatella** [comment]
I think the majority of requests that come through for this are just wanting something far simpler: transpose root, choose the corresponding “preset” diagram.  It’s actually likely to be more effective than a “smarter” algorithm, as transposing a “standard” G chord up a fifth would yield a “standard” D chord, even though that’s not a literal transposition of the original voicing at all.  Then maybe fall back to this sort algorithm for chords not part of that standard table.

**Comment by bkunda** [comment]
Agreed we need to solve this, but we'll be looking for the simplest solution possible (quite likely as @MarcSabatella suggested).

**Comment by MarcSabatella** [comment]
Something else that occurs to me for whenever someone looks at this:

As I said, not only is the table lookup simpler than a fancy algorithm, it's also more what most users actually will expect.  But also - it could be made customizable, so users can specify their own transposed versions of their favorite custom chords.  So, if you create a C7b9, you could also create Db7b9, D7b9, Eb7b9, etc, and enter them all into the table.  Once you've set all that up, then the transpose algorithm could then choose *your* D7b9 when you ask it to transpose C7b9 up a whole step.  And then only for the cases where the table is missing entries would we need a fancy algorithm to generate brand new fretboard diagrams.

So, the first pass at this might just have the built-in table so there is little design work required, but whoever implements it, might consider doing so in a way that would ultimately allows the table to be user-customized.

**Comment by zediiiii** [comment]
Proposed algorithm:

To get things started,

1. Make a library of basic **shiftable** chord fretboards based on chord spelling and root note (ie, all the major triads, minor triads, sus2 and sus4, 7 flavor chords). This guy has a great explanation on how to do that algorithmically instead of just one at a time. https://musescore.com/user/20856146/scores/6607649
2. Add in basic chords with **open** notes (same doc above lists all of these)
4. Chord analysis can be done on each nth beat, for example, on the quarter, on the half, or on the dotted quarter. The algorithm would shift to the next available note if that beat isn't available. Alternatively, the chord analysis could be done on every note.
5. Chord analysis would need to be able to span all staves and have the ability to exclude staves (passing tones might muddy the chords). It would also need to be able to be done by selection, not just for the whole piece.
6. Give the user a preference to prefer open chords or to prefer shiftable chords.
7. Users can select or change chords in a tree of choices (for example, C chords, then 7 chords, then Cmaj7).
8. An ideal solution would allow users to access a menu for each chord that provides possible voicings for the fretboard diagram. 

Time savers:
Use open-source code to generate the fretboard lists, ie https://github.com/hybridpicker/fretboard-position-finder
